pilot/user/atlashpc/common.py

```diff
@@ -0,0 +1,1692 @@
+        if version:
+            status = create_dbrelease(version, job.workdir)
+
+    # assign error in case of DBRelease handling failure
+    if not status:
+        job.piloterrorcodes, job.piloterrordiags = errors.add_error_code(errors.DBRELEASEFAILURE)
+
+    # cleanup job parameters if only copy-to-scratch
+    #if job.only_copy_to_scratch():
+    #    log.debug('job.params=%s' % job.jobparams)
+    #    if ' --usePFCTurl' in job.jobparams:
+    #        log.debug('cleaning up --usePFCTurl from job parameters since all input is copy-to-scratch')
+    #        job.jobparams = job.jobparams.replace(' --usePFCTurl', '')
+    #    if ' --directIn' in job.jobparams:
+    #        log.debug('cleaning up --directIn from job parameters since all input is copy-to-scratch')
+    #        job.jobparams = job.jobparams.replace(' --directIn', '')
+
+    return status
+
+
+def get_payload_command(job):
+    """
+    Return the full command for execuring the payload, including the sourcing of all setup files and setting of
+    environment variables.
+
+    :param job: job object.
+    :raises PilotException: TrfDownloadFailure.
+    :return: command (string).
+    """
+
+    log = get_logger(job.jobid)
+
+    # Should the pilot do the asetup or do the jobPars already contain the information?
+    prepareasetup = should_pilot_prepare_asetup(job.noexecstrcnv, job.jobparams)
+
+    # Get the platform value
+    # platform = job.infosys.queuedata.platform
+
+    # Is it a user job or not?
+    userjob = job.is_analysis()
+
+    # get the general setup command and then verify it
+    cmd = get_setup_command(job, prepareasetup)
+    ec, diagnostics = verify_setup_command(cmd)
+    if ec != 0:
+        raise PilotException(diagnostics, code=ec)
+
+    if is_standard_atlas_job(job.swrelease):
+
+        # Normal setup (production and user jobs)
+        log.info("preparing normal production/analysis job setup command")
+        cmd = get_normal_payload_command(cmd, job, prepareasetup, userjob)
+
+    else:  # Generic, non-ATLAS specific jobs, or at least a job with undefined swRelease
+
+        log.info("generic job (non-ATLAS specific or with undefined swRelease)")
+        cmd = get_generic_payload_command(cmd, job, prepareasetup, userjob)
+
+    # add any missing trailing ;
+    if not cmd.endswith(';'):
+        cmd += '; '
+    log.debug('post cmd: %s' % cmd)
+
+    # only if not using a user container
+    if not job.imagename:
+        site = os.environ.get('PILOT_SITENAME', '')
+        variables = get_payload_environment_variables(cmd, job.jobid, job.taskid, job.attemptnr, job.processingtype, site, userjob)
+        cmd = ''.join(variables) + cmd
+
+    cmd = cmd.replace(';;', ';')
+
+    # For direct access in prod jobs, we need to substitute the input file names with the corresponding TURLs
+    # get relevant file transfer info
+    use_copy_tool, use_direct_access, use_pfc_turl = get_file_transfer_info(job)
```

Comment:
Doug,

 you cloned files from outdated "next" branch, so that it does not include recent updates from merged requests, in particular #223 

In general, I fear that completely cloning `user/atlas` into `user/atlashpc` will be very difficult to support and maintain, since apart of the code duplication itself all updates need to be propagated twice which is a pain. it's better to extend (reuse) and overwrite only real hpc specific logic.

Reply:
Right, that would become too complicated (you told me at CHEP but I misunderstood you and thought it meant much fewer changes, I'm sorry). I suggest you clone the latest next branch - this is actually the most important part, but you might want to wait until the final patch which is pending, hopefully later today (BUILD 2.3.1 (28+)) - and then start with making pull requests for the changes you have here minus the user/atlashpc code. We can work out the changes you have there separately (since the code is duplicated I can't see which they are). Since we are currently focusing on running jobs at NERSC in the same way as on the grid, there shouldn't be too many differences to the generic workflow, so as long as we are running grid-like jobs on an HPC we should stick to that workflow.